cnd/src/btsieve/mod.rs

```diff
@@ -88,39 +88,37 @@ where
 /// This function yields all blocks as part of its process.
 /// This function returns the block-hashes of all visited blocks.
 async fn walk_back_until<C, P, B, H>(
-    stop_condition: P,
+    should_stop_here: P,
+    starting_block: B,
     connector: &C,
     co: &Co<B>,
-    block: B,
 ) -> anyhow::Result<HashSet<H>>
 where
     C: BlockByHash<Block = B, BlockHash = H>,
     P: Fn(&B) -> bool,
-    B: BlockHash<H> + PreviousBlockHash<H> + Clone,
+    B: BlockHash<H> + PreviousBlockHash<H>,
     H: Eq + Hash + Copy,
 {
-    let mut seen_blocks: HashSet<H> = HashSet::new();
-    let mut blockhash = block.block_hash();
+    let mut seen_blocks = HashSet::new();
 
-    co.yield_(block.clone()).await;
-    seen_blocks.insert(blockhash);
-
-    if stop_condition(&block) {
-        return Ok(seen_blocks);
-    } else {
-        blockhash = block.previous_block_hash();
-    }
+    let mut current_blockhash = starting_block.block_hash();
+    let mut current_block = starting_block;
 
     loop {
-        let block = connector.block_by_hash(blockhash).await?;
-        co.yield_(block.clone()).await;
-        seen_blocks.insert(blockhash);
+        seen_blocks.insert(current_blockhash);
 
-        if stop_condition(&block) {
+        // compute variables before consuming the block by yielding it
+        current_blockhash = current_block.previous_block_hash();
+        let should_stop_here = should_stop_here(&current_block);
+
+        // make sure we yield the block before we exit
+        co.yield_(current_block).await;
+
+        if should_stop_here {
             return Ok(seen_blocks);
-        } else {
-            blockhash = block.previous_block_hash();
         }
+
+        current_block = connector.block_by_hash(current_blockhash).await?
     }
```

Comment:
Its not immediately obvious from this diff as to why we don't have to call `previous_block_hash()`?  Just flagging it with this comment, leaving the judgement up to you.

Reply:
We don't have to because it is already computed before :)

Reply:
That is exactly what the comment above is supposed to explain but I failed apparently ..

Reply:
Writing clear documentation is non-trivial :)

Reply:
I pushed an update to the comment before it merged, have a look if that is more understandable!